src/tests/JIT/Directed/Convert/out_of_range_fp_to_int_conversions.cs

```diff
@@ -0,0 +1,448 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Numerics;
+
+namespace FPBehaviorApp
+{
+    public enum FPtoIntegerConversionType
+    {
+        CONVERT_BACKWARD_COMPATIBLE,
+        CONVERT_SENTINEL,
+        CONVERT_SATURATING,
+        CONVERT_NATIVECOMPILERBEHAVIOR,
+        CONVERT_MANAGED_BACKWARD_COMPATIBLE_X86_X64,
+        CONVERT_MANAGED_BACKWARD_COMPATIBLE_ARM32,
+    }
+
+    public enum ConversionType
+    {
+        ToInt,
+        ToLong,
+        ToUInt,
+        ToULong
+    }
+
+    public static class Native
+    {
+        [DllImport("out_of_range_fp_to_int_conversionsnative")]
+        [SuppressGCTransition]
+        public static extern int ConvertDoubleToLong(double x, FPtoIntegerConversionType t);
+
+        [DllImport("out_of_range_fp_to_int_conversionsnative")]
+        [SuppressGCTransition]
+        public static extern uint ConvertDoubleToUnsignedLong(double x, FPtoIntegerConversionType t);
+
+        [DllImport("out_of_range_fp_to_int_conversionsnative")]
+        [SuppressGCTransition]
+        public static extern long ConvertDoubleToLongLong(double x, FPtoIntegerConversionType t);
+
+        [DllImport("out_of_range_fp_to_int_conversionsnative")]
+        [SuppressGCTransition]
+        public static extern ulong ConvertDoubleToUnsignedLongLong(double x, FPtoIntegerConversionType t);
+    }
+
+    public static class Managed
+    {
+        public static void DumpBits(double d)
+        {
+            ulong uintVal = (ulong)BitConverter.DoubleToInt64Bits(d);
+            bool signBit = (int)((uintVal >> 63) & 0x1) != 0;
+            ulong mantissa = uintVal & 0x000F_FFFF_FFFF_FFFF;
+            int exponent = (int)((uintVal >> 52) & 0x7FF);
+            Console.WriteLine($"{signBit} {exponent:x} {mantissa:x}");
+        }
+
+        // Equivalent to Math.Truncate, but written in C# to avoid Debug/Check penalties running on CoreCLR test builds
+        public static double Truncate(double d)
+        {
+            ulong uintVal = (ulong)BitConverter.DoubleToInt64Bits(d);
+            int exponent = (int)((uintVal >> 52) & 0x7FF);
+            if (exponent < 1023)
+            {
+                uintVal &= 0x8000_0000_0000_0000ul;
+            }
+            else if (exponent < 1075)
+            {
+                uintVal &= (ulong)(~(0xF_FFFF_FFFF_FFFF >> (exponent - 1023)));
+            }
+
+            return BitConverter.Int64BitsToDouble((long)uintVal);
+        }
+
+        public static int ConvertDoubleToLong(double x, FPtoIntegerConversionType t)
+        {
+            if (t == FPtoIntegerConversionType.CONVERT_NATIVECOMPILERBEHAVIOR)
```

Comment:
Can we use `CONVERT_NATIVECOMPILERBEHAVIOR` on C# side? Do we expect to see the same result from C++ and C# in this test with `CONVERT_NATIVECOMPILERBEHAVIOR`?

Reply:
We can use it on both sides, but we do not expect the same results between C++ and C# in all cases. It really exists for a slightly larger set of experiments where I was looking at behavior differences between managed and native, and being able to request on both sides helped me understand what exactly was happening. I've left the logic in place, as we might want to re-run those experiments, and these functions are really convenient to use.

Reply:
ah, I was used to read "native compiler" as C++ compiler so that caused the confusion here. Thanks for the explanation.